### DISCUSSION TITLE
Improve `params` ergonomics for `getStaticPaths`

### DISCUSSION THREAD

**Discussion by natemoo-re** [withastro/roadmap#676] [Proposal]
### Body

# Summary

Currently, `getStaticPaths` has some behaviors with `params` that commonly trip users up.

We have [a complete recipe](https://docs.astro.build/en/core-concepts/routing/#example-dynamic-pages-at-multiple-levels) in docs to alleviate some confusion (`slug: undefined`), but maybe we can fix the API!

# Background & Motivation

It would be nice if `getStaticPaths` could support `params: { slug: '' }` or `params: { slug: 'index' }` to flatten a route rather than `params: { slug: undefined }`.

# Goals

- Provide better ergonomics for params coming from external data sources (a database or file probably won't encode `undefined`, for example)
- Reduce user confusion
- Make `getStaticPaths` easier to work with

# Example

```diff
export function getStaticPaths() {
  return data.map((content) => ({
-    params: { main: !content.slug ? undefined : content.slug },
+    params: { main: content.slug }, // supports '' or 'index'
    props: { content },
  }));
}
```

**Comment by isaac-mcfadyen** [comment]
This sounds great to me!

One thing: I'd discourage allowing `index`. There are many reasons a user might have a legitimate page called `index` (blog post, page signifying "index of the site", etc) and having it automatically assume `index` means "flatten this route" would cause confusion/tricky to debug issues.

An empty param makes it clear to me that it flattens the route (there's no param so there's nothing in that path section).

**Reply by jasikpark** [reply]
Could have an Astro Symbol() for "index" instead?

**Reply by delucis** [reply]
Agree with this that `index` is probably a bad idea. I can’t think of drawbacks for `''` though? 🤔
